Make it easy to build and test LXD

simondeziel opened the pull request:
Here what the workflow look like:

```
$ lxc launch ubuntu-minimal-daily:24.04 v1 --vm -p lxd-test
$ lxc shell v1
root@v1:~# cd lxd && make deps && make && make test-shell
LXD-TEST root@v1:~/lxd/test$ 
```

tomponline commented on test/main.sh:
> if that dir is backed by `virtiofs`, the tests fail.

How so?

simondeziel replied:
Yeah, I need to fill a bug for Aleks to have a proper look. He already expressed interest in fixing this behavior which IMHO will probably also help other use cases.

That said, having `tmp.XXX` dirs littering our local git clone still feels wrong to me.

tomponline replied:
good idea, we should do that for sure as it speeds things up a lot

simondeziel replied:
`LXD_TMPFS=1` avoids the `virtiofs` issue. However, even if `LXD_TMPFS=1` is desirable, I don't feel like `make test-shell` should set this environment variable, it has to be a conscious decision IMHO.

WDYT?

simondeziel replied:
We enable it in GH CI too so should we default to having this flag enabled? I still believe `test-shell` shouldn't add much magic on its own.

tomponline replied:
Can you add it to bash profile in cloud-init?

simondeziel replied:
I can and will do that but I'm still wondering why not change the default at this point ;)

simondeziel replied:
Ah shoot, I just realized that using `environment.LXD_TMPFS` would break when using `lxc shell` as that starts with clean env :/

simondeziel replied:
NVM, will use the same trick as in `lxd-ci`'s profile:

```
    runcmd:
    - echo "LXD_TMPFS=1" >> /etc/environment
```

tomponline commented on test/main.sh:
```diff
@@ -176,7 +176,7 @@ TEST_DIR="$(mktemp -d -t lxd-test.tmp.XXXX)"
 
 # Verify the dir chain is accessible for other users
-INACCESSIBLE_DIRS="$(namei -m "${TEST_DIR}" | awk '/^ d/ {print $1}' | grep -v 'x$' || true)"
+INACCESSIBLE_DIRS="$(namei -m "${TEST_DIR}" | awk '/^ d/ {print $1}' | grep -v '[xt]$' || true)"
```
What is this fixing? Can we add a comment explaining the cryptic command please?

simondeziel replied:
This is mostly a guard rail for if the `sudo chmod +x ~` is not done.

tomponline commented on test/main.sh:
```diff
@@ -273,13 +273,22 @@ if [ -n "${GITHUB_ACTIONS:-}" ]; then
   exit
 fi
 
+if [ "${1:-"all"}" = "tshell" ]; then
```
What is this and why do we need it? Commit message is lacking explanation.

simondeziel replied:
It gives you an interactive "test shell session". I augmented the commit messages and added comments in the script and `Makefile`.

tomponline replied:
can we call it `make testshell` to be clearer?

simondeziel replied:
Naming is hard. It started as `tash` for `test + bash`, then felt that `tshell` was more descriptive while still being short but I guess we go all in with explicit and use `test-shell`. I added the `-` as our other target separate words with it already.

tomponline commented on Makefile:
```diff
@@ -175,6 +175,11 @@ deps: dqlite liblxc
 	@echo "# Please set the following in your environment (possibly ~/.bashrc)"
 	@make -s env
 
+.PHONY: tshell
```
What is a tshell?

tomponline commented on doc/lxd-test.yaml:
```diff
@@ -0,0 +1,102 @@
+      commands:
+      - snap install go --classic
+  limits.cpu: "4"
+  limits.memory: 16GiB
```
This seems excessive and many users will have to reduce this (myself included!)

simondeziel replied:
This is to align with CPU/memory specs of GH action runners.

Would you like me to update the `sed` command doing the mangling to also tweak those 2 limits based on user provided variables?

simondeziel commented on CONTRIBUTING.md:
```diff
@@ -132,6 +132,51 @@ If you add or update configuration options, regenerate and commit the documentat
+lxc launch ubuntu-minimal-daily:24.04 v1 --vm -p lxd-test
+sleep 30
+# this may take a while as many packages need to be installed
+lxc exec v1 -- cloud-init status --wait --long
```
I'd rather not as that adds an extra step to be typed by the user. This whole PR is aiming for the easiest/quickest way to get a throw away env that is 100% clean. I use that multiple times a day so every bit count IMHO.

tomponline replied:
Ack

kadinsayani commented on CONTRIBUTING.md:
```diff
@@ -132,6 +132,51 @@ If you add or update configuration options, regenerate and commit the documentat
 
+#### Development environment setup
+
+Several pieces of software are needed in order to build and test LXD. Here is an easy way to create a virtual-machine to use as development environment. LXD itself is needed to power that virtual-machine so install it first: {ref}`installing`.
```
```suggestion
Several pieces of software are needed in order to build and test LXD. Here is an easy way to create a virtual-machine to use as a development environment. LXD itself is needed to power that virtual-machine so install it first: {ref}`installing`.
```

kadinsayani commented on CONTRIBUTING.md:
````diff
@@ -132,6 +132,51 @@ If you add or update configuration options, regenerate and commit the documentat
+
+Once LXD is installed and {ref}`initialized <initialize>`, a special profile (`lxd-test`) needs to be loaded:
+
+```sh
````
This `sh` block is empty.

kadinsayani commented:
What happens to the shell running lxd after the test exits (by ^C)? Will the normal cleanup steps happen?

simondeziel commented on doc/lxd-test.yaml:
```diff
@@ -0,0 +1,102 @@
+      # Speed things up by not pulling from backports/security and avoid restricted/multiverse pockets.
+      # In general, backported packages or those from restricted/multiverse shouldn't be relied on because
+      # they don't come with the same level of support as those from main for example.
+      # The security repo doesn't make much sense when pulling from a Canonical maintained archive mirror.
```
There is a lot to unpack here. Let's start by pointing out that there is a precedent in `lxd-ci.git`: https://github.com/canonical/lxd-ci?tab=readme-ov-file#lxd-ci-profile  (not that it's a good justification ;) )

So packages from `backports` or `multiverse` are not supported by Canonical, only `main` and `restricted` are (ignoring ESM/Pro). `restricted` is supported but is for proprietary software. In both cases, we don't need any of that as we only depend on bits from `main` and `universe` which are the 2 "pockets" not disabled.

As for why it's safe to `security` repo? Let's first start by why is `security.ubuntu.com` part of every default install. That is specifically to avoid a malicious apt mirror (country specific or other, like say `http://mirrors.dc.clear.net.ar/ubuntu/` to pick on the first one from list here https://launchpad.net/ubuntu/+archivemirrors). If that malicious mirror stops syncing with Ubuntu primary archives with the intention of serving old/stale/known vulnerable package, the apt client would still get the patched version from `security.ubuntu.com` that comes from Canonical directly.

Now in our case, `ubuntu:` images all pull packages from `archive.ubuntu.com` which is not a random mirror but the primary source of update, on par with `security.ubuntu.com` in terms of trust.

So all that to say that it's OK to "cut those corners" as we only save time, not reliability not security.

simondeziel commented on CONTRIBUTING.md:
````diff
@@ -132,6 +132,51 @@ If you add or update configuration options, regenerate and commit the documentat
+sed "s|@@PATH_TO_LXD_GIT@@|${GIT_ROOT}|" "${GIT_ROOT}/doc/lxd-test.yaml" | lxc profile create lxd-test
+```
+
+This profile can then be used to launch an Ubuntu Noble VM and start using it:
````
It has to be Noble. If you do go with 22.04 you miss the `yq` .deb package so you need to pull the snap.

If you go with a newer release, you are in uncharted territories. Should work but might run into some packages needing to be have their name updated.

tomponline commented:
@simondeziel please can you rebase too

skozina commented on doc/lxd-test.yaml:
```diff
@@ -0,0 +1,102 @@
+      path: /usr/local/bin/ci-wait
+      permissions: "0755"
+
+    runcmd:
```
Where did the `chmod -x /usr/sbin/lxc /usr/sbin/lxd` go? It is in your (excellent!) lxd-ci profile in lxd-ci.

skozina commented on doc/lxd-test.yaml:
```diff
@@ -0,0 +1,102 @@
+
+    package_update: true
+    package_upgrade: true
+    packages:
```
Can we also remove `uidmap` package as part of the setup? That was sometimes causing the uid remap errors.

skozina commented on doc/lxd-test.yaml:
```diff
@@ -0,0 +1,102 @@
+    - socat
+    - sqlite3
+    - swtpm
+    - yq
```
Bold ask -- can we include `vim` in the list, even if it's not needed?

kadinsayani replied:
+1 for this, can we also set `environment.EDITOR` to `vim`?

kadinsayani replied:
Or we can just inherit `EDITOR` from the host.

skozina reviewed (approved):
Few nits, I really appreciate we could build a standard test profile;-) I was using the one from lxd-ci since I found it... Thanks!

kadinsayani commented on CONTRIBUTING.md:
````diff
@@ -132,6 +132,56 @@ If you add or update configuration options, regenerate and commit the documentat
+
+```sh
+# enter the VM
+$ lxc exec v1 -- bash
````
```suggestion
lxc exec v1 -- bash
```

simondeziel replied:
I usually am with you in not including `$ ` as that makes copy pasting bloody annoying but I wanted to include the PS1 in those. Now looking back, I agree with you. Will drop them.

simondeziel commented:
@tomponline I think it's ready for a (final?) round of review. I've addressed all the feedback point and am waiting on this PR to be merged to report the `virtiofs` issue as this will simplify the reproducer.

tomponline reviewed (approved):
LGTM thanks!
